### PR TITLE
JVM IR: Get rid of FAKE_OWNER in MethodSignatureMapper

### PR DESCRIPTION
**Pull request by sfs**
Here's my understanding of why we have `FAKE_OWNER` in the first place. Please let my know if I missed something!

---

The JVM IR backend handles most intrinsics in lowerings or as part of `IrIntrinsicMethods`. However, there are nine intrinsics which the JVM backend only handles during inlining. These are the inline array constructors (`arrayOf`, `emptyArray`), three special coroutine intrinsics (`intercepted`, `coroutineContext`, `suspendCoroutineUninterceptedOrReturn`), the `typeOf` intrinsic, two generic enum methods (`enumValues`, `enumValueOf`), and the synthetic `alwaysEnabledAssert`.

Four of these look like regular methods before we enter the inliner (the coroutine intrinsics and `typeOf`). No special handling is necessary in `ExpressionCodegen`. Similarly, no special handling is necessary for `alwaysEnabledAssert`, since we only create and process calls to this method in the inliner itself.

The remaining four inline intrinsics really are intrinsics, however. There is no code for them in the standard library and *they do not have class parents* in the IR.

This means that `ExpressionCodegen` has to be careful when computing the owner for a function call, even after `IrIntrinsicMethods` have been resolved. In particular, `MethodSignatureMapper` puts in a `FAKE_OWNER` when it cannot resolve a call.

This PR lowers or implements the remaining inline intrinsics as regular `IrIntrinsicMethods` and adds more stringent error checking in `ExpressionCodegen`.

I think this is important, because right now there are cases where compilation succeeds while quietly producing incorrect bytecode. Case in point: There are 35 FIR tests which *passed* even though the resulting code contained calls to non-existent methods (which were never executed as part of the test).

### CODE REVIEW

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/EnumIntrinsics.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            // <REIFIED-OPERATIONS-MARKER>
+            // ICONST_0
+            // ANEWARRAY Ljava/lang/Enum;
+            putReifiedOperationMarkerIfTypeIsReifiedParameter(type, ReifiedTypeInliner.OperationKind.ENUM_REIFIED)
+            mv.iconst(0)
+            mv.newarray(AsmTypes.ENUM_TYPE)
+            MaterialValue(codegen, ENUM_ARRAY_TYPE, expression.type)
+        } else {
+            val enumType = typeMapper.mapType(type)
+            val enumArrayType = Type.getObjectType("[${enumType.descriptor}")
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/EnumIntrinsics.kt]: Minor suggestion to use `AsmUtil.getArrayType` here and below

File: compiler/testData/cli/jvm/firHello.args [hunk 2]
```diff
@@ -1,3 +1,4 @@
+// IGNORE_BACKEND: ANY
```
Comment (udalov) [inline compiler/testData/cli/jvm/firHello.args]: Maybe comment out the `kotlin.io.println` call instead, since the rest of the test works?

cc @demiurg906

**Comment by udalov** [comment]
Thank you! Your understanding is correct and this change looks great. Will merge after a remote run.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/EnumIntrinsics.kt (hunk 1 above)
Comment (sfs) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/EnumIntrinsics.kt]: Done!

File: compiler/testData/cli/jvm/firHello.args (hunk 2 above)
Comment (sfs) [reply compiler/testData/cli/jvm/firHello.args]: You're right, that makes more sense! Done.

**Comment by udalov** [comment]
Thanks, merged!

**Pull request merged**